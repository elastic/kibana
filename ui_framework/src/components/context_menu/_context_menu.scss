$kuiContextMenuWidth: $kuiSize * 16;

.kuiContextMenu {
  width: $kuiContextMenuWidth;
  position: relative;
  overflow: hidden;
  transition: height $kuiAnimSpeedFast $kuiAnimSlightResistance;
  border-radius: $kuiBorderRadius;

  .kuiContextMenu__content {
    padding: $kuiSizeS;
  }
}

  .kuiContextMenu__icon {
    margin-right: $kuiSizeS;
  }

  .kuiContextMenu__itemLayout {
    display: flex;
    align-items: center;
  }
