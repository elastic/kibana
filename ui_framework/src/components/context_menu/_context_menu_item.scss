.kuiContextMenuItem {
  padding: $kuiSizeM;
  width: 100%;
  text-align: left;
  color: $kuiTextColor;

  &:hover, &:focus {
    text-decoration: underline;
  }

  &:focus {
    background-color: $kuiFocusBackgroundColor;
  }
}

  .kuiContextMenuItem__inner {
    display: flex;
  }

  .kuiContextMenuItem__text {
    flex-grow: 1;
  }

  .kuiContextMenuItem__arrow {
    align-self: flex-end;
  }
