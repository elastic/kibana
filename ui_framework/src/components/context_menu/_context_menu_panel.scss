.kuiContextMenuPanel {
  position: absolute;
  width: 100%;
  visibility: visible;
  background-color: $kuiColorEmptyShade;

  &.kuiContextMenuPanel-txInLeft {
    pointer-events: none;
    animation: kuiContextMenuPanelTxInLeft $kuiAnimSpeedNormal $kuiAnimSlightResistance;
  }

  &.kuiContextMenuPanel-txOutLeft {
    pointer-events: none;
    animation: kuiContextMenuPanelTxOutLeft $kuiAnimSpeedNormal $kuiAnimSlightResistance;
  }

  &.kuiContextMenuPanel-txInRight {
    pointer-events: none;
    animation: kuiContextMenuPanelTxInRight $kuiAnimSpeedNormal $kuiAnimSlightResistance;
  }

  &.kuiContextMenuPanel-txOutRight {
    pointer-events: none;
    animation: kuiContextMenuPanelTxOutRight $kuiAnimSpeedNormal $kuiAnimSlightResistance;
  }
}

.kuiContextMenuPanel--next {
  transform: translateX($kuiContextMenuWidth);
  visibility: hidden;
}

.kuiContextMenuPanel--previous {
  transform: translateX(-$kuiContextMenuWidth);
  visibility: hidden;
}

.kuiContextMenuPanelTitle {
  width: 100%;
  text-align: left;
}

@keyframes kuiContextMenuPanelTxInLeft {
  0% {
    transform: translateX($kuiContextMenuWidth);
  }
  100% {
    transform: translateX(0);
  }
}

@keyframes kuiContextMenuPanelTxOutLeft {
  0% {
    transform: translateX(0);
  }
  100% {
    transform: translateX(-$kuiContextMenuWidth);
  }
}

@keyframes kuiContextMenuPanelTxInRight {
  0% {
    transform: translateX(-$kuiContextMenuWidth);;
  }
  100% {
    transform: translateX(0);
  }
}

@keyframes kuiContextMenuPanelTxOutRight {
  0% {
    transform: translateX(0);
  }
  100% {
    transform: translateX($kuiContextMenuWidth);
  }
}
