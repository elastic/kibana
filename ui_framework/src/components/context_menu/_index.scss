@import 'context_menu';
@import 'context_menu_panel';
@import 'context_menu_item';
