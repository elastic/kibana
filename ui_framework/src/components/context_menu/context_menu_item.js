import React, {
  cloneElement,
} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

export const KuiContextMenuItem = ({
  children,
  className,
  hasPanel,
  icon,
  ...rest,
}) => {
  let iconInstance;

  if (icon) {
    iconInstance = cloneElement(icon, {
      className: classNames(icon.props.className, 'kuiContextMenu__icon'),
    });
  }

  let arrow;

  if (hasPanel) {
    arrow = <span className="kuiContextMenu__arrow kuiIcon fa-angle-right" />;
  }

  const classes = classNames('kuiContextMenuItem', className);

  return (
    <button
      className={classes}
      {...rest}
    >
      <span className="kuiContextMenu__itemLayout">
        {iconInstance}
        <span className="kuiContextMenuItem__text">
          {children}
        </span>
        {arrow}
      </span>
    </button>
  );
};

KuiContextMenuItem.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  icon: PropTypes.element,
  onClick: PropTypes.func,
  hasPanel: PropTypes.bool,
};
