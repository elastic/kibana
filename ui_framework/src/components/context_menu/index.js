export {
  KuiContextMenu,
} from './context_menu';

export {
  KuiContextMenuPanel,
} from './context_menu_panel';

export {
  KuiContextMenuItem,
} from './context_menu_item';
