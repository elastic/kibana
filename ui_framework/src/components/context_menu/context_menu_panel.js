import React, {
  cloneElement,
  Component,
} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import tabbable from 'tabbable';

import { KuiPopoverTitle } from '../../components';
import { cascadingMenuKeyCodes } from '../../services';

const transitionDirectionAndTypeToClassNameMap = {
  next: {
    in: 'kuiContextMenuPanel-txInLeft',
    out: 'kuiContextMenuPanel-txOutLeft',
  },
  previous: {
    in: 'kuiContextMenuPanel-txInRight',
    out: 'kuiContextMenuPanel-txOutRight',
  },
};

export class KuiContextMenuPanel extends Component {
  static propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    title: PropTypes.string,
    onClose: PropTypes.func,
    onHeightChange: PropTypes.func,
    transitionType: PropTypes.oneOf(['in', 'out']),
    transitionDirection: PropTypes.oneOf(['next', 'previous']),
    onTransitionComplete: PropTypes.func,
    onUseKeyboardToNavigate: PropTypes.func,
    hasFocus: PropTypes.bool,
    items: PropTypes.array,
    showNextPanel: PropTypes.func,
    showPreviousPanel: PropTypes.func,
    initialFocusedItemIndex: PropTypes.number,
  }

  static defaultProps = {
    hasFocus: true,
    items: [],
  }

  constructor(props) {
    super(props);

    this.menuItems = [];
    this.state = {
      isTransitioning: Boolean(props.transitionType),
      focusedItemIndex: props.initialFocusedItemIndex,
    };
  }

  incrementFocusedItemIndex = amount => {
    let nextFocusedItemIndex;

    if (this.state.focusedItemIndex === undefined) {
      // If this is the beginning of the user's keyboard navigation of the menu, then we'll focus
      // either the first or last item.
      nextFocusedItemIndex = amount < 0 ? this.menuItems.length - 1 : 0;
    } else {
      nextFocusedItemIndex = this.state.focusedItemIndex + amount;

      if (nextFocusedItemIndex < 0) {
        nextFocusedItemIndex = this.menuItems.length - 1;
      } else if (nextFocusedItemIndex === this.menuItems.length) {
        nextFocusedItemIndex = 0;
      }
    }

    this.setState({
      focusedItemIndex: nextFocusedItemIndex,
    });
  };

  onKeyDown = e => {
    // If this panel contains items you can use the left arrow key to go back at any time.
    // But if it doesn't contain items, then you have to focus on the back button specifically,
    // since there could be content inside the panel which requires use of the left arrow key,
    // e.g. text inputs.
    if (
      this.props.items.length
      || document.activeElement === this.backButton
      || document.activeElement === this.panel
    ) {
      if (e.keyCode === cascadingMenuKeyCodes.LEFT) {
        if (this.props.showPreviousPanel) {
          this.props.showPreviousPanel();

          if (this.props.onUseKeyboardToNavigate) {
            this.props.onUseKeyboardToNavigate();
          }
        }
      }
    }

    if (this.props.items.length) {
      switch (e.keyCode) {
        case cascadingMenuKeyCodes.TAB:
          // We need to sync up with the user if s/he is tabbing through the items.
          const focusedItemIndex = this.menuItems.indexOf(document.activeElement);

          this.setState({
            focusedItemIndex:
              (focusedItemIndex >= 0 && focusedItemIndex < this.menuItems.length)
                ? focusedItemIndex
                : undefined,
          });
          break;

        case cascadingMenuKeyCodes.UP:
          e.preventDefault();
          this.incrementFocusedItemIndex(-1);

          if (this.props.onUseKeyboardToNavigate) {
            this.props.onUseKeyboardToNavigate();
          }
          break;

        case cascadingMenuKeyCodes.DOWN:
          e.preventDefault();
          this.incrementFocusedItemIndex(1);

          if (this.props.onUseKeyboardToNavigate) {
            this.props.onUseKeyboardToNavigate();
          }
          break;

        case cascadingMenuKeyCodes.RIGHT:
          if (this.props.showNextPanel) {
            e.preventDefault();
            this.props.showNextPanel(this.state.focusedItemIndex);

            if (this.props.onUseKeyboardToNavigate) {
              this.props.onUseKeyboardToNavigate();
            }
          }
          break;

        default:
          break;
      }
    }
  };

  updateFocus() {
    // If this panel has lost focus, then none of its content should be focused.
    if (!this.props.hasFocus) {
      if (this.panel.contains(document.activeElement)) {
        document.activeElement.blur();
      }
      return;
    }

    // Setting focus while transitioning causes the animation to glitch, so we have to wait
    // until it's finished before we focus anything.
    if (this.state.isTransitioning) {
      return;
    }

    // If there aren't any items then this is probably a form or something.
    if (!this.menuItems.length) {
      // If we've already focused on something inside the panel, everything's fine.
      if (this.panel.contains(document.activeElement)) {
        return;
      }

      // Otherwise let's focus the first tabbable item and expedite input from the user.
      if (this.content) {
        const tabbableItems = tabbable(this.content);
        if (tabbableItems.length) {
          tabbableItems[0].focus();
        }
      }
      return;
    }

    // If an item is focused, focus it.
    if (this.state.focusedItemIndex !== undefined) {
      this.menuItems[this.state.focusedItemIndex].focus();
      return;
    }

    // Focus on the panel as a last resort.
    if (!this.panel.contains(document.activeElement)) {
      this.panel.focus();
    }
  }

  onTransitionComplete = () => {
    this.setState({
      isTransitioning: false,
    });

    if (this.props.onTransitionComplete) {
      this.props.onTransitionComplete();
    }
  }

  componentDidMount() {
    this.updateFocus();
  }

  componentWillReceiveProps(nextProps) {
    // Clear refs to menuItems if we're getting new ones.
    if (nextProps.items !== this.props.items) {
      this.menuItems = [];
    }

    if (nextProps.transitionType) {
      this.setState({
        isTransitioning: true,
      });
    }
  }

  componentDidUpdate() {
    this.updateFocus();
  }

  menuItemRef = (index, node) => {
    // There's a weird bug where if you navigate to a panel without items, then this callback
    // is still invoked, so we have to do a truthiness check.
    if (node) {
      // Store all menu items.
      this.menuItems[index] = node;
    }
  };

  panelRef = node => {
    this.panel = node;

    if (this.panel) {
      if (this.props.onHeightChange) {
        this.props.onHeightChange(this.panel.clientHeight);
      }
    }
  };

  contentRef = node => {
    this.content = node;
  };

  render() {
    const {
      children,
      className,
      onClose,
      title,
      onHeightChange, // eslint-disable-line no-unused-vars
      transitionType,
      transitionDirection,
      onTransitionComplete, // eslint-disable-line no-unused-vars
      onUseKeyboardToNavigate, // eslint-disable-line no-unused-vars
      hasFocus, // eslint-disable-line no-unused-vars
      items,
      initialFocusedItemIndex, // eslint-disable-line no-unused-vars
      showNextPanel, // eslint-disable-line no-unused-vars
      showPreviousPanel, // eslint-disable-line no-unused-vars
      ...rest
    } = this.props;
    let panelTitle;

    if (title) {
      if (Boolean(onClose)) {
        panelTitle = (
          <button
            className="kuiContextMenuPanelTitle"
            onClick={onClose}
            ref={node => { this.backButton = node; }}
            data-test-subj="contextMenuPanelTitleButton"
          >
            <span className="kuiContextMenu__itemLayout">
              <span className="kuiContextMenu__icon kuiIcon fa-angle-left" />
              <span className="kuiContextMenu__text">
                {title}
              </span>
            </span>
          </button>
        );
      } else {
        panelTitle = (
          <KuiPopoverTitle>
            <span className="kuiContextMenu__itemLayout">
              {title}
            </span>
          </KuiPopoverTitle>
        );
      }
    }

    const classes = classNames('kuiContextMenuPanel', className, (
      this.state.isTransitioning && transitionDirectionAndTypeToClassNameMap[transitionDirection]
        ? transitionDirectionAndTypeToClassNameMap[transitionDirection][transitionType]
        : undefined
    ));

    const content = items.length
      ? items.map((MenuItem, index) => cloneElement(MenuItem, {
        buttonRef: this.menuItemRef.bind(this, index),
      }))
      : children;

    return (
      <div
        ref={this.panelRef}
        className={classes}
        onKeyDown={this.onKeyDown}
        tabIndex="0"
        onAnimationEnd={this.onTransitionComplete}
        {...rest}
      >
        {panelTitle}

        <div ref={this.contentRef}>
          {content}
        </div>
      </div>
    );
  }
}
