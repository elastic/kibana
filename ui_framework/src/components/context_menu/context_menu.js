import React, {
  Component,
} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { KuiContextMenuPanel } from './context_menu_panel';
import { KuiContextMenuItem } from './context_menu_item';

export class KuiContextMenu extends Component {
  static propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    initialPanelId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    idToPanelMap: PropTypes.object,
    idToPreviousPanelIdMap: PropTypes.object,
  }

  static defaultProps = {
    idToPanelMap: {},
    idToPreviousPanelIdMap: {},
  }

  constructor(props) {
    super(props);

    this.resetTransitionId = undefined;

    this.state = {
      outGoingPanelId: undefined,
      currentPanelId: props.initialPanelId,
      transitionDirection: undefined,
    };
  }

  showPanel(panelId, direction) {
    clearTimeout(this.resetTransitionId);

    this.setState({
      outGoingPanelId: this.state.currentPanelId,
      currentPanelId: panelId,
      transitionDirection: direction,
    });

    // Queue the transition to reset.
    this.resetTransitionId = setTimeout(() => {
      this.setState({
        transitionDirection: undefined,
      });
    }, 250);
  }

  updateHeight() {
    const height = this.currentPanel.clientHeight;
    this.menu.setAttribute('style', `height: ${height}px`);
  }

  renderPanel(panelId, transitionType) {
    const panel = this.props.idToPanelMap[panelId];

    if (!panel) {
      return;
    }


    const renderItems = items => items.map(item => {
      let onClick;

      if (item.onClick) {
        onClick = item.onClick;
      } else if (item.panel) {
        onClick = this.showPanel.bind(this, item.panel.id, 'next');
      }

      return (
        <KuiContextMenuItem
          key={item.name}
          icon={item.icon}
          onClick={onClick}
          hasPanel={Boolean(item.panel)}
        >
          {item.name}
        </KuiContextMenuItem>
      );
    });

    const previousPanelId = this.props.idToPreviousPanelIdMap[panelId];

    let onClose;

    if (typeof previousPanelId === 'number') {
      onClose = this.showPanel.bind(this, previousPanelId, 'previous');
    }

    return (
      <KuiContextMenuPanel
        panelRef={node => { this.currentPanel = node; }}
        title={panel.title}
        onClose={onClose}
        transitionType={transitionType}
        transitionDirection={this.state.transitionDirection}
      >
        {panel.content || renderItems(panel.items)}
      </KuiContextMenuPanel>
    );
  }

  componentDidMount() {
    this.updateHeight();
  }

  componentDidUpdate() {
    this.updateHeight();
  }

  render() {
    const {
      idToPanelMap, // eslint-disable-line no-unused-vars
      idToPreviousPanelIdMap, // eslint-disable-line no-unused-vars
      className,
      initialPanelId, // eslint-disable-line no-unused-vars
      ...rest,
    } = this.props;

    const currentPanel = this.renderPanel(this.state.currentPanelId, 'in');
    const outGoingPanel = this.renderPanel(this.state.outGoingPanelId, 'out');

    const classes = classNames('kuiContextMenu', className);

    return (
      <div
        ref={node => { this.menu = node; }}
        className={classes}
        {...rest}
      >
        {outGoingPanel}
        {currentPanel}
      </div>
    );
  }
}
