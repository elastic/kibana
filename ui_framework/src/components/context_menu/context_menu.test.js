import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test/required_props';

import { KuiContextMenu } from './context_menu';

describe('KuiContextMenu', () => {
  test('is rendered', () => {
    const component = render(
      <KuiContextMenu {...requiredProps} />
    );

    expect(component)
      .toMatchSnapshot();
  });
});
