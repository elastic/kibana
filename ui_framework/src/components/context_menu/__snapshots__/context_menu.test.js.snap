// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KuiContextMenu is rendered 1`] = `
<div
  aria-label="aria-label"
  class="kuiContextMenu testClass1 testClass2"
  data-test-subj="test subject string"
/>
`;
